File: Aditya Vardhan Singh/Other Problems/Day 12/328 Odd Even Linked List/C++/OddEvenLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        // If list has not more than 3 nodes, return list
        if (!(head && head->next && head->next->next)) {
            return head;
        }

        // Separate odd and even (indexed) lists
        ListNode * oddHead = head;
        ListNode * evenHead = head->next;

        ListNode *oddTemp = oddHead;
        ListNode *evenTemp = evenHead;

        while (oddTemp->next || evenTemp->next) {
            if (evenTemp->next) {
                oddTemp->next = evenTemp->next;
                evenTemp->next = nullptr;
                oddTemp = oddTemp->next;
            }
            if (oddTemp->next) {
                evenTemp->next = oddTemp->next;
                oddTemp->next = nullptr;
                evenTemp = evenTemp->next;
            }
        }

        // Join odd and even list
        oddTemp->next = evenHead;

        return oddHead;
    }
};

File: Aditya Vardhan Singh/Other Problems/Day 12/328 Odd Even Linked List/Python/OddEvenLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # If list has not more than 3 nodes, return list
        if not head or not head.next or not head.next.next:
            return head

        # Separate odd and even (index) list
        oddHead = head
        evenHead = head.next

        oddTemp = oddHead
        evenTemp = evenHead

        while oddTemp.next or evenTemp.next:
            if evenTemp.next:
                oddTemp.next = evenTemp.next
                evenTemp.next = None
                oddTemp = oddTemp.next
            if oddTemp.next:
                evenTemp.next = oddTemp.next
                oddTemp.next = None
                evenTemp = evenTemp.next
        
        oddTemp.next = evenHead

        return oddHead
